ygen/protogen.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package ygen
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/openconfig/goyang/pkg/yang"
+)
+
+// protoMsgField describes a field of a protobuf message.
+type protoMsgField struct {
+	Tag        uint32            // Tag is the field number that should be used in the protobuf message.
+	Name       string            // Name is the field's name.
+	Type       string            // Type is the protobuf type for the field.
+	IsRepeated bool              // IsRepeated indicates whether the field is repeated.
+	Extensions map[string]string // Extensions is the set of field tags that are applied to the field.
+}
+
+// protoMsg describes a protobuf message.
+type protoMsg struct {
+	Name     string           // Name is the name of the protobuf message to be output.
+	YANGPath string           // YANGPath stores the path that the message corresponds to within the YANG schema.
+	Fields   []*protoMsgField // Fields is a slice of the fields that are within the message.
+}
+
+// genProtoMsg takes an input yangStruct which describes a container or list entry
+// within the YANG schema and returns a protoMsg which can be mapped to the protobuf
+// code representing it. It uses the set of messages that have been extracted and the
+// current generator state to map to other messages and ensure uniqueness of names.
+func genProtoMsg(msg *yangStruct, msgs map[string]*yangStruct, state *genState) (protoMsg, []error) {
+	var errs []error
+
+	msgDef := protoMsg{
+		// msg.name is already specified to be CamelCase in the form we expect it
+		// to be for the protobuf message name.
+		Name:     msg.name,
+		YANGPath: slicePathToString(msg.path),
+	}
+
+	definedFieldNames := map[string]bool{}
+
+	for name, field := range msg.fields {
+		fieldDef := &protoMsgField{
+			Name: makeNameUnique(safeProtoFieldName(name), definedFieldNames),
+		}
+
+		t, err := protoTagForEntry(field)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("proto: could not generate tag for field %s: %v", field.Name, err))
+		}
+		fieldDef.Tag = t
+
+		switch {
+		case field.IsList():
+			errs = append(errs, fmt.Errorf("proto: list generation unimplemented for %s", field.Path()))
+			continue
```

Comment:
Suggest creating var err error at the top of the loop, then set it when you hit error inside here. Then check for error and append at the exit to the switch. This will remove the need to continue on error.

Reply:
Done.